From e53e42a5cfee8ae17f75e3d7e4533ce98aabeab4 Mon Sep 17 00:00:00 2001
From: Mingcong Bai <jeffbai@aosc.io>
Date: Mon, 27 Jan 2025 04:58:49 +0800
Subject: [PATCH 4/5] AOSCOS: dracut-init: decompress kernel modules and
 firmware before final compression

Decompress all kernel modules and firmware (if compressed), this allows
for better compression ratio as we make the compressed initramfs.

This may save over 10MiB post-XZ.

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 dracut-init.sh | 48 ++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 48 insertions(+)

diff --git a/dracut-init.sh b/dracut-init.sh
index 1ce0d7ed..914ab3e2 100755
--- a/dracut-init.sh
+++ b/dracut-init.sh
@@ -1029,6 +1029,54 @@ for_each_module_dir() {
 dracut_kernel_post() {
     local dstdir="${dstdir:-"$initdir"}"
 
+    # Check for the necessary if(1) command for post-processing.
+    if ! find_binary find &> /dev/null; then
+        derror "Will not process kernel modules and firmware as find(1) is not available!"
+        return 1
+    fi
+
+    # Check for necessary compression utilities.
+    for cmd in gzip xz zstd; do
+        if ! find_binary "$cmd" &> /dev/null; then
+            derror "Will not process kernel modules and firmware compressed with '$cmd' as the command is not available!"
+            return 1
+        fi
+    done
+
+    # Decompress all firmware and module data for more optimal final
+    # compression ratio.
+    for kerndir in "${dstdir}/lib/firmware" \
+                   "${dstdir}/lib/modules/$kernel"; do
+        for kernfile in `find "$kerndir" -type f`; do
+            case "$kernfile" in
+                *.gz)
+                    ddebug "Decompressing gzipped kernel module/firmware: $kernfile ..."
+                    gzip -d "$kernfile"
+                    ;;
+                *.xz)
+                    ddebug "Decompressing xz-compressed kernel module/firmware: $kernfile ..."
+                    xz -d "$kernfile"
+                    ;;
+                *.zst)
+                    ddebug "Decompressing zstd-compressed kernel module/firmware: $kernfile ..."
+                    zstd -q --rm -d "$kernfile"
+                    ;;
+            esac
+        done
+    done
+
+    # Fixup symlinks to compressed firmware.
+    for fwlink in `find "${dstdir}/lib/firmware" \
+                       -name '*.gz' -o -name '*.xz' -o -name '*.zst' \
+                       -type l`; do
+        ddebug "Fixing up decompressed firmware: $fwlink ..."
+        fwtarget=$(readlink $fwlink)
+        rm $fwlink
+        for ext in gz xz zst; do
+            ln -s ${fwtarget%.*} ${fwlink%.*} && break
+        done
+    done
+
     for _f in modules.builtin modules.builtin.alias modules.builtin.modinfo modules.order; do
         [[ -e $srcmods/$_f ]] && inst_simple "$srcmods/$_f" "/lib/modules/$kernel/$_f"
     done
-- 
2.48.1

